### PR TITLE
Hide personal collections and items when configuring click behavior

### PR DESCRIPTION
**Pull request by WiNloSt**
This is the final part of the epic https://github.com/metabase/metabase/issues/34733

This PR implements
- Hide personal collections and dashboards when configuring click behavior linking to dashboards
- Hide personal collections and questions when configuring click behavior linking to questions

image

#### How to verify

Dashboard click behavior

1. Link to questions/dashboards for a question in a public collection
    **expectation**:
    - show only questions/dashboards in public collections
    - search should return only questions/dashboards in public collections
1. Link to questions/dashboards for a question in a personal collection
    **expectation**:
    - show all questions/dashboards
    - search should return all questions/dashboards

### CODE REVIEW

**Comment by WiNloSt** [comment]
Current dependencies on/for this PR:
* master
  * **PR #35135** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35135" target="_blank">Graphite</a> 
    * **PR #35141** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35141" target="_blank">Graphite</a> 
      * **PR #35142** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35142" target="_blank">Graphite</a> 
        * **PR #35171** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35171" target="_blank">Graphite</a> 
          * **PR #35213** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35213" target="_blank">Graphite</a> 
            * **PR #35214** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35214" target="_blank">Graphite</a> 
              * **PR #35215** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35215" target="_blank">Graphite</a> 
                * **PR #35277** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35277" target="_blank">Graphite</a> 
                  * **PR #35317** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35317" target="_blank">Graphite</a>  👈
    * **PR #35137** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35137" target="_blank">Graphite</a> 
      * **PR #35145** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35145" target="_blank">Graphite</a> 
      * **PR #35138** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35138" target="_blank">Graphite</a> 
        * **PR #35139** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35139" target="_blank">Graphite</a> 
    * **PR #35136** <a href="https://app.graphite.dev/github/pr/metabase/metabase/35136" target="_blank">Graphite</a> 

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/metabase/metabase/35317?utm_source=stack-comment).

File: frontend/src/metabase/collections/containers/FormCollectionPicker/FormCollectionPicker.tsx [hunk 1]
```diff
@@ -107,7 +108,8 @@ function FormCollectionPicker({
 
   const isOpenCollectionInPersonalCollection = openCollection?.is_personal;
   const showCreateNewCollectionOption =
-    !showOnlyPersonalCollections || isOpenCollectionInPersonalCollection;
+    filterPersonalCollections !== "only" ||
```
Comment (WiNloSt) [inline frontend/src/metabase/collections/containers/FormCollectionPicker/FormCollectionPicker.tsx:111]: The previous condition was `!showOnlyPersonalCollections`

Given `showOnlyPersonalCollections`  === `filterPersonalCollections === "only"`

that means `!showOnlyPersonalCollections` === `filterPersonalCollections !== "only"`

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.tsx [hunk 2]
```diff
@@ -92,7 +92,9 @@ const AddToDashSelectDashModal = ({
   if (shouldCreateDashboard) {
     return (
       <CreateDashboardModal
-        showOnlyPersonalCollections={isQuestionInPersonalCollection}
+        filterPersonalCollections={
+          isQuestionInPersonalCollection ? "only" : undefined
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.tsx:96]: `undefine` means showing all collections

File: frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx [hunk 3]
```diff
@@ -86,7 +87,7 @@ function ItemPicker<TId>({
   className,
   showSearch = true,
   showScroll = true,
-  showOnlyPersonalCollections = false,
+  filterPersonalCollections,
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx:90]: We can remove the default value because the default is now `undefined` means we'll show everything from both public collections and personal collections.

File: frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx [hunk 4]
```diff
@@ -123,18 +124,14 @@ function ItemPicker<TId>({
 
     const collectionItems = list
       .filter(canWriteToCollectionOrChildren)
-      .filter(
-        showOnlyPersonalCollections
-          ? collection => collection.is_personal
-          : _.identity,
-      )
+      .filter(getCollectionFilter(filterPersonalCollections))
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx:127]: a new function is defined below

File: frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.tsx [hunk 5]
```diff
@@ -178,11 +182,11 @@ function LinkedEntityPicker({
     });
   }, [clickBehavior, updateSettings]);
 
-  const { data: dashboard } = useDashboardQuery({
+  const { data: targetDashboard } = useDashboardQuery({
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.tsx:185]: Rename this to make it obvious that this is the selected dashboard as a click behavior target. So it's not confused with `dashboard` (introduced below) which is the dashboard that we're configuring the click behavior for.

File: frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.unit.spec.tsx [hunk 6]
```diff
@@ -0,0 +1,382 @@
...
+            models: ["card", "dataset"],
+            q: typedText,
+          });
+        });
+      });
+    });
+  });
+});
+
+function fromEntriesWithMultipleValues(
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.unit.spec.tsx:365]: The reason this one is needed is because previously I was comparing the `URLSearchParams` with:
https://github.com/metabase/metabase/blob/68788d775f5642046630ad34c98c4c5e0a5e9522/frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx#L888-L891

This doesn't work with some of the tests in this file because it has multiple value for the same key i.e. `models=dataset&models=card`

using `Object.fromEntries` will just result in an object with the key `models` being either `dataset` or `card`.

This little util produces the value array, when it found multiple value for the same key.

Comment (WiNloSt) [inline frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.unit.spec.tsx:365]: The reason this one is needed is because previously I was comparing the `URLSearchParams` with:
https://github.com/metabase/metabase/blob/68788d775f5642046630ad34c98c4c5e0a5e9522/frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.unit.spec.tsx#L888-L891

This doesn't work with some of the tests in this file because it has multiple value for the same key i.e. `models=dataset&models=card`

using `Object.fromEntries` will just result in an object with the key `models` being either `dataset` or `card`.

This little util produces the value array, when it found multiple value for the same key.

File: frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx [hunk 7]
```diff
@@ -150,19 +147,17 @@ function ItemPicker<TId>({
...
+        query.filter_items_in_personal_collection = filterPersonalCollections;
       }
     } else {
       query.collection = openCollectionId;
     }
 
-    if (models.length === 1) {
-      query.models = models;
-    }
+    query.models = models;
```
Comment (WiNloSt) [inline frontend/src/metabase/containers/ItemPicker/ItemPicker.tsx:157]: This is caught during unit tests 😉 . It was kind of a bug that we fetche all results instead of just the specified models. There is one case where we set `models = ["dataset", "card"]` and we didn't apply `models` param at all.
#### After the fix
Screenshot 2023-11-09 at 7 36 58 PM

#### Before the fix
Screenshot 2023-11-09 at 7 37 01 PM


Although we used to fetch results from all models it doesn't make any difference in `ItemPicker` since FE will filter those results based on `models` anyway.

https://github.com/metabase/metabase/blob/68788d775f5642046630ad34c98c4c5e0a5e9522/frontend/src/metabase/containers/ItemPicker/ItemPickerView.tsx#L185

tl;dr: This change is needed for unit tests and as a result it fetches only necessary data from the API.

File: frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.tsx [hunk 8]
```diff
@@ -233,6 +237,12 @@ function LinkedEntityPicker({
     ],
   );
 
+  const dashboard = useSelector(getDashboard);
```
Comment (npretto) [inline frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.tsx:240]: Maybe worth naming this `currentDashboard`? I know the other dashboard is called `targetDashboard` but considering before this change, dashboard meant the target one it could make sense

File: frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.unit.spec.tsx (hunk 6 above)
Comment (npretto) [reply frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.unit.spec.tsx:365]: There's also `url.getAll` which could simplify the code if you want:
```js
const urlSearchParamsToObject = (searchParams) => {
  const obj = {};
  for (const [key] of Array.from(searchParams)) {
    const value = searchParams.getAll(key)
    obj[key] = value.length > 1 ? value : value[0];
  }
  return obj;
};
```

**Review by deniskaber — APPROVED** [review]
Added several inline comments

**Review by cdeweyx — APPROVED** [review]

File: frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.tsx (hunk 8 above)
Comment (WiNloSt) [reply frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.tsx:240]: We use `dashboard` pretty much as a value returned from `getDashboard` everywhere, so I'd like to keep that consistency.

Also, usually with `currentSomething` it should be used to differentiate from other Something e.g. `previousDashboard` vs `currentDashboard` if we need to compare the 2 versions for some reasons. But that's not the case here either.

File: frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.tsx [hunk 9]
```diff
@@ -289,5 +300,8 @@ function LinkedEntityPicker({
   );
 }
 
-// eslint-disable-next-line import/no-default-export -- deprecated usage
-export default LinkedEntityPicker;
+function isPublicCollection(collection: Partial<Collection>) {
```
[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.tsx:303]: Yeah, I'll make that into a util since it's used pretty much everywhere now. 🙏

File: frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.unit.spec.tsx (hunk 6 above)
Comment (WiNloSt) [reply frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.unit.spec.tsx:365]: Fixed in 395e9d020150ca4621a2848135b2ff3c8419d500

File: frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.tsx (hunk 2 above)
Comment (WiNloSt) [reply frontend/src/metabase/containers/AddToDashSelectDashModal/AddToDashSelectDashModal.tsx:96]: This is intended to match the query we send to the search endpoints `filter_items_in_personal_collection: "only" | "exclude"`

Although its nuance is a little different because in the search endpoints we're filtering the results, which in this case are only dashboards, questions, or datasets (models)

`filterPersonalCollections: "only" | "exclude" | undefined` are meant not only for filtering items, but also collections that are shown in any picker UI too. 

#### collections
Screenshot 2023-11-10 at 2 23 31 PM

#### search results
Screenshot 2023-11-10 at 2 24 51 PM

Do you have any suggestion on improving this flag name?

File: frontend/src/metabase-lib/parameters/utils/click-behavior.ts [hunk 10]
```diff
@@ -391,7 +392,9 @@ function getParameter(
     (clickBehavior.linkType === "dashboard" ||
       clickBehavior.linkType === "question")
   ) {
-    const dashboard = (extraData.dashboards || {})[clickBehavior.targetId];
+    const dashboard = (extraData.dashboards || {})[
```
[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase-lib/parameters/utils/click-behavior.ts]: Fixed in 9c525a9f8a2626ff2bba5b21aa01ead33115a5d4

File: frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/dashboard/components/ClickBehaviorSidebar/LinkOptions/LinkedEntityPicker/LinkedEntityPicker.tsx]: Fixed in c95c2dac57b144a04ecfba84fcda767aa205b2d9

File: frontend/src/metabase/collections/utils.ts [hunk 11]
```diff
@@ -11,15 +11,27 @@ export function nonPersonalOrArchivedCollection(
...
 
-export function isPersonalCollection(
+export function isRootPersonalCollection(
   collection: Partial<Collection> | CollectionItem,
 ): boolean {
   return typeof collection.personal_owner_id === "number";
 }
 
+export function isPersonalCollection(
+  collection: Pick<Collection, "is_personal">,
```
Comment (WiNloSt) [inline frontend/src/metabase/collections/utils.ts:24]: We want to allow passing `ROOT_COLLECTION` directly and since it doesn't contain all the necessary properties in `Collection`, we can't use `Collection` directly.

And this util only care if an object contains `is_personal`, so this type should be enough.

File: frontend/src/metabase/dashboard/selectors/selectors.js [hunk 12]
```diff
@@ -77,6 +77,9 @@ export const getIsShowDashboardInfoSidebar = createSelector(
   sidebar => sidebar.name === SIDEBAR_NAME.info,
 );
 
+/**
+ * @type {(state: import("metabase-types/store").State) => import("metabase-types/api").Dashboard}
```
Comment (WiNloSt) [inline frontend/src/metabase/dashboard/selectors/selectors.js:81]: This ensures TypeScript files using `getDashboard` will get the correct type without having to convert this whole file to TypeScript which is not trivial

File: frontend/src/metabase/lib/urls/collections.ts [hunk 13]
```diff
@@ -40,8 +41,7 @@ export function collection(collection?: Collection) {
     return `/collection/${id}`;
   }
 
-  const isPersonalCollection = typeof collection.personal_owner_id === "number";
-  const slug = isPersonalCollection
+  const slug = isRootPersonalCollection(collection)
```
Comment (WiNloSt) [inline frontend/src/metabase/lib/urls/collections.ts:44]: This was implemented exactly as this function's implementation

https://github.com/metabase/metabase/blob/a57007a03a6a9ba9172f950311fc6cb2828982b3/frontend/src/metabase/collections/utils.ts#L17-L21
